perf: optimize spectator retrieval and add unit tests (gtest)

dudantas opened the pull request:
Improves performance and correctness of spectator retrieval in Map and Tile by tracking creature ranges and providing efficient access methods. Refactors Thing API for safer null handling, updates the Lua binding for const member functions, and adds GTest-based unit tests for spectator logic. CMake and vcpkg have been updated to support test builds and the Google Test (GTest) dependency.

<img width="1428" height="616" alt="image" src="https://github.com/user-attachments/assets/d9e92bfb-99a7-4d3a-a773-382d90be3d13" />

This pull request refactors the CMake build system to introduce a new `otclient_core` static library, separating core source files from platform-specific entry points. It also improves build configuration flexibility, consolidates build flags, and modernizes C++ code by making type-checking methods `const`. Additionally, some Lua bindings are updated to handle overloaded methods correctly, and minor code hygiene improvements are made in the client codebase.

**Build System Refactor and Improvements**

* Introduced a new `otclient_core` static library in `src/CMakeLists.txt`, moving most source files to this target and linking it from the main executable. Platform-specific entry points (`main.cpp` and `androidmain.cpp`) are now added conditionally. This improves modularity and build flexibility. [[1]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L371-L373) [[2]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R390-R476)
* Moved build flags, precompiled header, unity build, and sanitizer configuration to apply to `otclient_core`, and consolidated these options for consistency across platforms. [[1]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L78-L142) [[2]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R390-R476)
* Updated include directories and linked libraries to use `otclient_core` instead of the main target, ensuring correct dependency resolution for all platforms (Windows, Linux, Android, WASM). [[1]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L488-R513) [[2]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L509-R527) [[3]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L539-R557) [[4]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L555-R573) [[5]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L589-R607) [[6]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L603-R621) [[7]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L647-R664) [[8]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L673-R691) [[9]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L688-R706) [[10]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L728-R753)
* Changed Discord RPC build flag handling to use a CMake variable (`OTCLIENT_ENABLE_DISCORD_RPC`) and propagate it as a preprocessor definition to `otclient_core`. [[1]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L216-R160) [[2]](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R390-R476)

**C++ Code Modernization**

* Updated type-checking and status methods (`isCreature`, `isNpc`, `isMonster`, `isItem`, `isContainer`, `isEffect`, `isLocalPlayer`) in `creature.h`, `item.h`, `effect.h`, and `localplayer.h` to be `const` for better const-correctness and API clarity. [[1]](diffhunk://#diff-3bac30eb8bceb4f9bfe01a0bd42cb00e923e72d7a08150e3abe4a1c768407d5bR151-R157) [[2]](diffhunk://#diff-3bac30eb8bceb4f9bfe01a0bd42cb00e923e72d7a08150e3abe4a1c768407d5bL360-R372) [[3]](diffhunk://#diff-33cc6d175cb90672884a3342c4915fc7f8b9dda0c185955a13e14894fbd7ddd6L36-R36) [[4]](diffhunk://#diff-a3c5fe745a17a8def687bd3117c359adc43a5d04fdcbcc39a5b282149fb8916fL112-R112) [[5]](diffhunk://#diff-a3c5fe745a17a8def687bd3117c359adc43a5d04fdcbcc39a5b282149fb8916fL151-R151) [[6]](diffhunk://#diff-42b8bc70a913a9403330b887102fc8d98ff09d3cbd1117f2ed0ae78530254a8aL131-R131)
* Added new silent setter methods (`setOldPositionSilently`, `setRemovedSilently`) and a const getter (`getOldPosition`) to `Creature` for improved state management. [[1]](diffhunk://#diff-3bac30eb8bceb4f9bfe01a0bd42cb00e923e72d7a08150e3abe4a1c768407d5bR211-R213) [[2]](diffhunk://#diff-3bac30eb8bceb4f9bfe01a0bd42cb00e923e72d7a08150e3abe4a1c768407d5bR151-R157)

**Lua Bindings Update**

* Updated Lua bindings in `luafunctions.cpp` to use explicit overload resolution for methods with both const and non-const versions (`isRemoved`, `hasCreatures`). This prevents ambiguity and ensures correct method exposure to Lua scripts. [[1]](diffhunk://#diff-962b1fe1290388640199de520e0c3ef65c88299cafd659a297a23373f181b514L614-R617) [[2]](diffhunk://#diff-962b1fe1290388640199de520e0c3ef65c88299cafd659a297a23373f181b514L936-R942)

**Code Hygiene and Minor Improvements**

* Improved header inclusion style in `map.cpp` for consistency and added a helper function `cleanNewSpectators` for maintaining unique spectator lists.
* Optimized memory usage in `Map::getSpectatorsInRangeEx` by reserving container sizes and using an unordered set for faster lookup.

**Testing Infrastructure**

* Added an option (`OTCLIENT_BUILD_TESTS`) to enable or disable building unit tests via CMake, improving build customization for development and CI.

dudantas committed:
perf: optimize spectator retrieval and add unit tests (gtest)

Improves performance and correctness of spectator retrieval in Map and Tile by tracking creature ranges and providing efficient access methods. Refactors Thing API for safer null handling, updates Lua binding for const member functions, and adds GTest-based unit tests for spectator logic. CMake and vcpkg updated to support test builds and GTest dependency.

dudantas committed:
Add range filtering test for map spectators

Introduces a new test, RangeFiltering, to verify that getSpectatorsInRangeEx correctly filters spectators based on specified range parameters. The test covers both near and far range scenarios to ensure proper functionality.

dudantas committed:
Ensure unique and ordered spectators in map queries

Added deduplication of creatures in Map::getSpectatorsInRangeEx and Map::getSpectatorsByPattern using an unordered_set. Updated Tile::appendSpectators to filter by position. Extended tests to verify deterministic ordering, uniqueness, and correct creature span updates on addition and removal.

dudantas committed:
Add virtual destructor to ApplicationContext and update tests

Added a virtual destructor to ApplicationContext for proper cleanup in derived classes. Updated map_spectators_test.cpp to use try_emplace instead of emplace for m_knownCreatures to avoid overwriting existing entries.

libergod commented:
@dudantas Ready to submit? 
<img width="2048" height="686" alt="image" src="https://github.com/user-attachments/assets/d283d146-efed-4d6b-b2bb-06b702461fff" />

dudantas committed:
Remove position check in appendSpectators

Eliminated the redundant position check for creatures in Tile::appendSpectators, allowing all creatures on the tile to be appended regardless of their position.

dudantas committed:
Refactor Lua class registration templates

Replaces direct calls to stdext::demangle_class with local variables in template methods for class registration and member/static function binding. This improves code readability and maintainability.

dudantas committed:
Make type-checking methods const and refactor spectators logic

Changed type-checking methods (e.g., isItem, isCreature) to be const across Thing and derived classes for better const-correctness. Refactored Map's getSpectatorsInRangeEx and getSpectatorsByPattern to use a helper for cleaning duplicate/new spectators. Updated Thing to make most property accessors const and improved code clarity. Replaced std::reverse with std::ranges::reverse in Tile::getCreatures.

dudantas commented:
> @dudantas Ready to submit?

Fixed, test again.

libergod commented:
Still having warnings about luainterface

<img width="1963" height="626" alt="image" src="https://github.com/user-attachments/assets/1fbd5250-2063-4283-b96f-ef28dc07df0d" />

libergod commented on src/framework/luaengine/luabinder.h:
```diff
@@ -169,8 +169,10 @@ namespace luabinder
-                throw LuaException("failed to call a member function because the passed object is nil");
+            if (!obj) {
+                g_logger.warning("Lua warning: member function call skipped because the passed object is nil");
+                return Ret{};
```
If I change this back, luabinder stops giving warnings:
<img width="1875" height="507" alt="image" src="https://github.com/user-attachments/assets/7c04c989-41b0-449c-a6f5-b7fa1bdfb98e" />

libergod commented on src/framework/luaengine/luabinder.h:
```diff
@@ -233,4 +237,47 @@ namespace luabinder
+        return [=](const std::shared_ptr<C>& obj, const Args&... args) mutable -> Ret {
+            if (!obj) {
+                g_logger.warning("Lua warning: member function call skipped because the passed object is nil");
+                return Ret{};
```
fix used in both cases:
<img width="1068" height="521" alt="image" src="https://github.com/user-attachments/assets/e5a33659-50af-408f-a289-f5da11a4ed8c" />

libergod reviewed (commented):
Needed to do some changes to fix luabinder warnings, can you check @dudantas ?

dudantas committed:
Add void return type support to make_mem_func helpers

Enhanced make_mem_func and make_mem_func_singleton to correctly handle member functions with void return types using if constexpr and overloads. This prevents returning a default value for void functions and improves Lua binding robustness.

dudantas committed:
Merge remote-tracking branch 'upstream/main' into perf-spectators-map

dudantas committed:
Add null checks to Thing property accessors

Refactored Thing class methods to safely handle cases where getThingType() returns nullptr, preventing potential crashes. Default values are now returned for all property accessors if the underlying type is missing.

dudantas committed:
Reorder and update includes in global.h and test

Moved 'using namespace std::literals;' in global.h to follow include statements for better style. Added missing 'thingtype.h' include and adjusted macro undef order in map_spectators_test.cpp to improve test reliability and maintainability.

dudantas committed:
Add getThingType and terminateWalk to DummyCreature

Implemented the getThingType and terminateWalk methods in the DummyCreature test class to provide required overrides for testing. This improves test coverage and ensures compatibility with interfaces expecting these methods.

dudantas commented:
Please retest this @libergod

dudantas committed:
Set default vcpkg SHA and remove pin validation

Workflows now use a default vcpkg commit SHA if LUKKA_RUN_VCPKG_SHA is unset, improving reliability. Removed redundant steps that validated the SHA variable, simplifying workflow configuration.

libergod committed:
Squashed commit of the following:

commit d22a2f9b78b6ad4ba742ca1f3986e775f100b21d
Author: Luan Luciano <luanluciano@outlook.com>
Date:   Fri Nov 14 17:07:44 2025 -0300

    fix: memory leak that occurs with each call to the decrypt() function

    Allocation of new uint8_t[size] without delete[] before the return

commit 8cbb23341292206e0617d64c9d394669c3da6210
Author: Eduardo Dantas <eduardo.dantas@hotmail.com.br>
Date:   Fri Nov 14 13:58:06 2025 -0300

    ci/cl: set default vcpkg SHA and remove pin validation (#1442)

commit 0232575afdc3fc7ae91f98e33583ab1f9291693f
Author: Eduardo Dantas <eduardo.dantas@hotmail.com.br>
Date:   Thu Nov 13 16:23:03 2025 -0300

    fix: add virtual destructor to ApplicationContext (#1441)

libergod commented:
Same warnings from luabinder

<img width="1590" height="675" alt="image" src="https://github.com/user-attachments/assets/be7f87ae-91c2-44e8-8cd9-fc09fef41862" />

dudantas committed:
Add make_default_return_value for safe default returns

Introduced the make_default_return_value template to provide safe default return values, especially for reference types. Updated member function and function binding logic to use this helper instead of direct value initialization.

libergod committed:
Fix warnings

libergod commented:
100%

libergod reviewed (approved):
100%

dudantas committed:
Merge remote-tracking branch 'upstream/main' into perf-spectators-map
